pkg/validator/cluster.go

```diff
@@ -832,6 +833,64 @@ func validateNamingScheme(
 	return patches
 }
 
+// normalizeResourceQuantities is called by normalizeRolesResourceQuantities
+// to handle each role resource list. If the quantity of a resource is
+// equivalent between old and new lists, then copy the new value into the old
+// list. If any resources have been removed or added, this function will
+// bail out early returning false; otherwise it returns true once done.
+func normalizeResourceQuantities(
+	prevList corev1.ResourceList,
+	newList corev1.ResourceList,
+) bool {
+	if len(newList) != len(prevList) {
+		return false
+	}
+	for res, newQuant := range newList {
+		oldQuant, ok := prevList[res]
+		if !ok {
+			return false
+		}
+		if newQuant.Cmp(oldQuant) == 0 {
+			prevList[res] = newQuant
+		}
+	}
+	return true
+}
+
+// normalizeRolesResourceQuantities makes sure that quantitatively-equivalent
+// resource values don't cause the "spec is changing" check to trigger.
+// (Sometimes the format of these quantities gets changed behind our backs.)
+// If the quantity of a resource is equivalent between old and new role specs,
+// then copy the new value into the old specs. Note that if the list of roles
+// is reordered, or if any roles or resources have been removed or added, this
+// function will return early.
+func normalizeRolesResourceQuantities(
+	prevRoles []kdv1.Role,
+	newRoles []kdv1.Role,
+) {
+	numRoles := len(prevRoles)
+	if len(newRoles) != numRoles {
+		return
+	}
+	for i := 0; i < numRoles; i++ {
+		prev := &(prevRoles[i])
+		new := &(newRoles[i])
+		if prev.Name != new.Name {
+			return
+		}
+		prevLimits := prev.Resources.Limits
+		newLimits := new.Resources.Limits
+		if !normalizeResourceQuantities(prevLimits, newLimits) {
+			return
+		}
+		prevRequests := prev.Resources.Requests
+		newRequests := new.Resources.Requests
+		if !normalizeResourceQuantities(prevRequests, newRequests) {
+			return
+		}
```

Comment:
Wouldn't it be easier to make equality function for ResourceList and then simply assigning prev.Resources.XXX to new.Resources.XXX?

Reply:
Maybe! Hmm. I think the equality function would have to look pretty much like the current code, just without the assignment that is currently happening when the equality is detected. So I'm not sure it would result in less code.